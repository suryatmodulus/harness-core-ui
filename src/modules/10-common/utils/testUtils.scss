/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

.browser {
  border: 1px solid #cccccc;
  border-radius: 8px;
  min-height: 300px;
  overflow: auto;

  .browser-header {
    background-color: #cccccc;
    padding: 8px 16px;
    display: flex;
  }

  .browser-path {
    display: block;
    border: none;
    line-height: 14px;
    font-size: 12px;
    color: #444444;
    border-radius: 9px;
    min-width: 100px;
    padding: 2px 8px;
    flex: 1 1 auto;
    border: 1px solid transparent;

    &:focus {
      border-color: rgb(30, 167, 253);
      outline: none;
    }
  }

  .browser-content {
    padding: 8px;
  }
}
