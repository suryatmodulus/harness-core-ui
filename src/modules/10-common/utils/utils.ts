/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { IconName } from '@harness/uicore'
import { Scope } from '@common/interfaces/SecretsInterface'
import { ModuleName } from 'framework/types/ModuleName'

interface SetPageNumberProps {
  setPage: (value: React.SetStateAction<number>) => void
  pageItemsCount?: number
  page: number
}

export const getModuleIcon = (module: ModuleName): IconName => {
  switch (module) {
    case ModuleName.CD:
      return 'cd-main'
    case ModuleName.CV:
      return 'cv-main'
    case ModuleName.CI:
      return 'ci-main'
    case ModuleName.CE:
      return 'ce-main'
    case ModuleName.CF:
      return 'cf-main'
  }
  return 'nav-project'
}

export const getReference = (scope?: Scope, identifier?: string): string | undefined => {
  switch (scope) {
    case Scope.PROJECT:
      return identifier
    case Scope.ORG:
      return `org.${identifier}`
    case Scope.ACCOUNT:
      return `account.${identifier}`
  }
}

export const setPageNumber = ({ setPage, pageItemsCount, page }: SetPageNumberProps): void => {
  if (pageItemsCount === 0 && page > 0) {
    setPage(page - 1)
  }
}

export const formatCount = (num: number): string => {
  const min = 1e3
  if (num >= min) {
    const units = ['k', 'M', 'B', 'T']
    const order = Math.floor(Math.log(num) / Math.log(1000))
    const finalNum = Math.round(num / 1000 ** order)
    if (finalNum === min && order < units.length) {
      return 1 + units[order]
    }
    return finalNum + units[order - 1]
  }
  return num.toLocaleString()
}
