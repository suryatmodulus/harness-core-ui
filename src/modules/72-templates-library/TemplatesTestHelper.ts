/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { ResponseString, ResponsePageTemplateSummaryResponse } from 'services/template-ng'
import type { UseGetMockDataWithMutateAndRefetch } from '@common/utils/testUtils'

export const stageMockTemplatesInputYaml: ResponseString = {
  status: 'SUCCESS',
  data:
    'type: "Deployment"' +
    '\nspec:' +
    '\n  serviceConfig:' +
    '\n    serviceRef: "<+input>"' +
    '\n  infrastructure:' +
    '\n    infrastructureDefinition:' +
    '\n      type: "KubernetesDirect"' +
    '\n      spec:' +
    '\n        namespace: "<+input>"' +
    '\n  execution:' +
    '\n    steps:' +
    '\n    - step:' +
    '\n        identifier: "Step_1"' +
    '\n        type: "ShellScript"' +
    '\n        spec:' +
    '\n          source:' +
    '\n            type: "Inline"' +
    '\n            spec:' +
    '\n              script: "<+input>"' +
    '\n'
}

export const mockTemplatesInputYaml: ResponseString = {
  status: 'SUCCESS',
  data:
    'type: "HarnessApproval"\ntimeout: "<+input>"' +
    '\nspec:\n  approvalMessage: "<+input>"\n  approvers:\n    userGroups: "<+input>"\n    minimumCount: "<+input>"\n'
}

export const mockTemplates: ResponsePageTemplateSummaryResponse = {
  status: 'SUCCESS',
  data: {
    content: [
      {
        accountId: 'kmpySmUISimoRrJL6NL73w',
        orgIdentifier: 'default',
        projectIdentifier: 'Templateproject',
        identifier: 'manjutesttemplate',
        name: 'manju-test-template-qq-12344',
        description:
          'Flink is a versatile framework, supporting many different deployment scenarios in a mix and match fashion.',
        tags: { QAR: '', 'Internal 1': '', 'Canary 1': '', BLUE: '', 'Tag A': '' },
        yaml:
          'template:' +
          '\n    name: manju-test-template-qq-12344' +
          '\n    identifier: manjutesttemplate' +
          '\n    versionLabel: v4' +
          '\n    type: Step' +
          '\n    projectIdentifier: Templateproject' +
          '\n    orgIdentifier: default' +
          '\n    description: Flink is a versatile framework, supporting many different deployment scenarios in a mix and match fashion.' +
          '\n    tags:' +
          '\n        QAR: ""' +
          '\n        Internal 1: ""' +
          '\n        Canary 1: ""' +
          '\n        BLUE: ""' +
          '\n        Tag A: ""' +
          '\n    spec:' +
          '\n        type: HarnessApproval' +
          '\n        timeout: <+input>' +
          '\n        spec:' +
          '\n            approvalMessage: <+input>' +
          '\n            includePipelineExecutionHistory: true' +
          '\n            approvers:' +
          '\n                userGroups: <+input>' +
          '\n                minimumCount: <+input>' +
          '\n                disallowPipelineExecutor: false' +
          '\n            approverInputs:' +
          '\n                - name: "1"' +
          '\n                  defaultValue: "1"' +
          '\n                - name: ttt' +
          '\n                  defaultValue: ttt' +
          '\n        failureStrategies:' +
          '\n            - onFailure:' +
          '\n                  errors:' +
          '\n                      - Timeout' +
          '\n                  action:' +
          '\n                      type: MarkAsSuccess' +
          '\n',
        versionLabel: 'v4',
        templateEntityType: 'Step',
        childType: 'HarnessApproval',
        templateScope: 'project',
        version: 8,
        gitDetails: {},
        lastUpdatedAt: 1635626311830,
        stableTemplate: true
      }
    ],
    pageable: {
      sort: { sorted: true, unsorted: false, empty: false },
      pageSize: 20,
      pageNumber: 0,
      offset: 0,
      paged: true,
      unpaged: false
    },
    last: true,
    totalPages: 1,
    totalElements: 1,
    sort: { sorted: true, unsorted: false, empty: false },
    first: true,
    numberOfElements: 5,
    size: 20,
    number: 0,
    empty: false
  }
}

export const mockTemplatesSuccessResponse: UseGetMockDataWithMutateAndRefetch<ResponsePageTemplateSummaryResponse> = {
  loading: false,
  refetch: jest.fn(),
  mutate: jest.fn(),
  cancel: jest.fn(),
  data: mockTemplates
}
