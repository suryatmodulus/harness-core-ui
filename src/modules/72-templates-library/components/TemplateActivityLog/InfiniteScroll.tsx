/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { MutableRefObject, useCallback, useEffect, useRef, useState } from 'react'

const DEFAULT_PAGE_SIZE = 10

interface GetItemsWithOffset {
  offset?: number
  limit?: number
}

interface InfiniteScrollProps {
  // getItems is a function promise that will resolve to an array of items to be displayed
  getItems: (options: GetItemsWithOffset) => Promise<any>

  /*
  ref of the element which when visible in the DOM shall trigger the next call
  to bring the next set of data. Check the usage in TemplateActivityLog
  */
  loadMoreRef: MutableRefObject<Element | null>

  // limit prop for defining the size of pagination, default is DEFAULT_PAGE_SIZE
  limit?: number
}

interface InfiniteScrollReturnProps {
  fetching: boolean
  error: string
  items: any
  hasMore: MutableRefObject<boolean>
  offsetToFetch: MutableRefObject<number>
  loadItems: () => void
  attachRefToLastElement: (index: number) => boolean
}

/*
Hook for infinite scroll

It will only work with NG APIs that support pagination and resolve with this format
{ content: [], empty: false, pageIndex: 1, pageItemCount: 3, pageSize: 10, totalItems: 13, totalPages: 2 }

What will this hook manage?
 - Items that are visible on the view. Can be used with any list of items that are resulted from paginated API calls

loadItems
 - Fetch data based on a given offset and limit
 - Function that accepts a page number/offset variable. It will call the underlying API function with the offset variable
*/
export const useInfiniteScroll = (props: InfiniteScrollProps): InfiniteScrollReturnProps => {
  const { getItems, limit = DEFAULT_PAGE_SIZE } = props
  const [items, setItems] = useState<any>([])
  const [fetching, setFetching] = useState(false)
  const [error, setError] = useState('')
  const hasMore = useRef(false)

  const initialPageLoaded = useRef(false)
  const offsetToFetch = useRef(0)

  const loadItems = () => {
    setFetching(true)

    getItems({
      offset: offsetToFetch.current,
      limit
    })
      .then(response => {
        setFetching(false)

        // If the cuurent fetch count exceeds totalItems, set hasMore as false
        const canFetchMore =
          response.data.totalItems > response.data.pageIndex * response.data.pageSize + response.data.pageItemCount
        hasMore.current = canFetchMore

        const responseContent = response.data.content
        setItems((prevItems: any) => [...prevItems, ...responseContent])
        setError('')
      })
      .catch(err => {
        setFetching(false)
        setError(err)
        setItems([])
      })
  }

  /*
  returns true if index is the last element of the list
  this is done so that the last element can be noticed by the IntersectionObserver
  */
  const attachRefToLastElement = useCallback(
    (index: number) => {
      return index === items.length - 1 && hasMore.current && !fetching
    },
    [fetching, hasMore.current]
  )

  /*
  call loadItems() and fetch the next batch of data if the target is now visible in viewport
  hasMore is true
  fetching is false

  Example - the last element of the list. We identify the last element by applying ref in the above callback
  */
  const loadMoreCallback = useCallback(
    entries => {
      const target = entries[0]
      if (target.isIntersecting && hasMore.current && !fetching) {
        offsetToFetch.current += 1
        loadItems()
      }
    },
    [fetching, hasMore.current, loadItems]
  )

  // Set the intersection observer
  useEffect(() => {
    // Default options for the IO
    // 0.25 threshold means call the callback function when 25% of the target is visible in the DOM
    const options = {
      threshold: 0.25
    }

    const observer = new IntersectionObserver(loadMoreCallback, options)

    // loadMoreRef is the target element received from props
    if (props.loadMoreRef && props.loadMoreRef.current) {
      observer.observe(props.loadMoreRef.current)
    }

    return () => {
      props.loadMoreRef?.current && observer.unobserve(props.loadMoreRef?.current)
    }
  }, [props.loadMoreRef, loadMoreCallback])

  // Just to ensure we don't end up fetching multiple times in the initial load
  useEffect(() => {
    if (initialPageLoaded.current) {
      return
    }

    loadItems()
    initialPageLoaded.current = true
  }, [loadItems])

  return {
    fetching,
    error,
    items,
    hasMore,
    loadItems,
    attachRefToLastElement,
    offsetToFetch
  }
}
