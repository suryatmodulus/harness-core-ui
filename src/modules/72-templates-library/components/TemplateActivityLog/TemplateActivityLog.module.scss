/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

.container {
  position: relative;
  .activity {
    display: flex;
    margin: var(--spacing-medium) 0;
  }

  .timelineSeparator {
    width: 0.2px;
    border: 0.2px dashed var(--form-field-border);
    height: 65px;
    margin: -5px 0px -25px 13px;
  }

  .timelineSeparatorLast {
    height: 25px;
    margin: -5px 0px 0px 13px;
  }

  .dot {
    height: 9px;
    width: 9px;
    margin-left: 9px;
    background-color: var(--form-field-border);
    border-radius: 50%;
    display: inline-block;
  }

  .lastIcon {
    margin-left: 6px !important;
  }

  .timelineFirstIcon {
    background-color: var(--primary-5) !important;
    color: white !important;
    padding: var(--spacing-2) !important;
    min-height: var(--spacing-6) !important;
    height: var(--spacing-6) !important;
    margin: var(--spacing-2) 0 0 var(--spacing-2) !important;
    min-width: var(--spacing-6) !important;
    box-shadow: none;
  }
}
