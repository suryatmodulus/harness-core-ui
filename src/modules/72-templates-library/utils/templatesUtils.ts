import type { UseStringsReturn } from 'framework/strings'

export enum TemplateType {
  Step = 'Step',
  Stage = 'Stage',
  Pipeline = 'Pipeline',
  Service = 'Service',
  Infrastructure = 'Infrastructure',
  StepGroup = 'StepGroup',
  Execution = 'Execution'
}

export const AllTemplatesTypes = 'All'

export interface TemplateTypeOption {
  label: string
  value: string
  disabled?: boolean
}

export const getAllowedTemplateTypes = (getString: UseStringsReturn['getString']): TemplateTypeOption[] => [
  {
    label: getString('step'),
    value: TemplateType.Step
  },
  {
    label: getString('templatesLibrary.stageTemplate'),
    value: TemplateType.Stage
  },
  {
    label: getString('common.pipeline'),
    value: TemplateType.Pipeline,
    disabled: true
  },
  {
    label: getString('service'),
    value: TemplateType.Service,
    disabled: true
  },
  {
    label: getString('infrastructureText'),
    value: TemplateType.Infrastructure,
    disabled: true
  },
  {
    label: getString('stepGroup'),
    value: TemplateType.StepGroup,
    disabled: true
  },
  {
    label: getString('executionText'),
    value: TemplateType.Execution,
    disabled: true
  }
]
