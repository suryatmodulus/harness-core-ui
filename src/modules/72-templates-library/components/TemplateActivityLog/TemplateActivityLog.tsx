/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React, { useCallback, useMemo, useRef } from 'react'
import cx from 'classnames'
import ReactTimeago from 'react-timeago'
import { Button, Color, Container, Heading, Icon, Layout, PageError, Text } from '@wings-software/uicore'
import { isEmpty } from 'lodash-es'
import { PageSpinner } from '@common/components'
import { useStrings } from 'framework/strings'
import { AuditEventDTO, AuditFilterProperties, getAuditEventListPromise } from 'services/audit'
import type { TemplateSummaryResponse } from 'services/template-ng'

import { useInfiniteScroll } from './InfiniteScroll'
import css from './TemplateActivityLog.module.scss'

export interface TemplateActivityLogProps {
  template: TemplateSummaryResponse
}

interface TemplateActivityProps {
  auditEvent: AuditEventDTO
  last: boolean
}

const HumanizedAction = ({ action }: { action: string }) => {
  /*
  Convert an action enum to human readable text string
  E.g. CREATE mapped to created, UPDATE to updated, etc
  */
  const { getString } = useStrings()
  let actionString = ''
  if (action === 'CREATE') {
    actionString = getString('templatesLibrary.created')
  } else if (action === 'UPDATE') {
    actionString = getString('templatesLibrary.updated')
  } else if (action === 'DEFAULT') {
    actionString = getString('templatesLibrary.default')
  }
  return <Text>{actionString}</Text>
}

const TimelineIndicators = ({ last }: { last: boolean }) => {
  /*
  The icons on the left pane of timeline
  Any change in this component should be sync'd with the right side data too
  */
  return (
    <div>
      <Button iconProps={{ size: 10 }} withoutBoxShadow round className={css.timelineFirstIcon} icon="edit" />
      <div className={css.timelineSeparator}></div>
      <div className={css.dot} />
      <div className={cx(css.timelineSeparator, last ? css.timelineSeparatorLast : '')}></div>
      {last ? <Icon name="double-chevron-down" color={Color.GREY_250} className={css.lastIcon} /> : null}
    </div>
  )
}

const TemplateActivity = ({ auditEvent, last }: TemplateActivityProps) => {
  /* Rendering a single audit activity */
  const { action, authenticationInfo, resource, timestamp, auditEventData } = auditEvent
  const { labels } = authenticationInfo
  const { getString } = useStrings()

  return (
    <div className={css.activity}>
      <TimelineIndicators last={last} />
      <div>
        <Layout.Horizontal spacing="xsmall" margin={{ left: 'small', top: 'xsmall' }}>
          <Text intent="primary">{labels?.username || getString('common.userLabel')}</Text>
          <HumanizedAction action={action} />
          <Text intent="primary">{resource.labels?.versionLabel || ''}</Text>
        </Layout.Horizontal>

        <Text font={{ size: 'small' }} margin={{ left: 'medium' }}>
          <ReactTimeago date={timestamp} />
        </Text>

        <Container
          padding="small"
          margin={{ left: 'medium', top: 'medium' }}
          border={{ radius: 4, color: Color.PRIMARY_6 }}
        >
          <Text font={{ size: 'small' }} width={420} lineClamp={2}>
            {(auditEventData as any)?.comments || getString('templatesLibrary.sampleTemplateAuditComment')}
          </Text>
        </Container>
      </div>
    </div>
  )
}

export const TemplateActivityLog = (props: TemplateActivityLogProps) => {
  const {
    template: { accountId }
  } = props
  const { getString } = useStrings()

  /* Ref to attach to the last element to the list - for infinite scroll */
  const loadMoreRef = useRef(null)

  /* The request body for the activity logs POST call */
  const postCallRequestBody: AuditFilterProperties = useMemo(() => {
    return {
      filterType: 'Audit',
      scopes: [
        {
          accountIdentifier: props.template.accountId,
          orgIdentifier: props.template.orgIdentifier,
          projectIdentifier: props.template.projectIdentifier
        }
      ],
      resources: [
        {
          type: 'TEMPLATE',
          identifier: props.template.identifier || ''
        }
      ],
      modules: ['TEMPLATESERVICE']
    }
  }, [props])

  const {
    items: templateActivityLogs,
    error: templateActivityLogsFetchError,
    fetching: fetchingTemplateActivityLogs,
    attachRefToLastElement,
    offsetToFetch
  } = useInfiniteScroll({
    getItems: options => {
      return getAuditEventListPromise({
        queryParams: { accountIdentifier: accountId || '', pageIndex: options.offset, pageSize: options.limit },
        body: postCallRequestBody
      })
    },
    limit: 10,
    loadMoreRef
  })

  const isfetchingLogsFirstTime = useCallback(() => {
    return fetchingTemplateActivityLogs && offsetToFetch.current === 0
  }, [fetchingTemplateActivityLogs, offsetToFetch.current])

  const isfetchingLogsNextTime = useCallback(() => {
    return fetchingTemplateActivityLogs && offsetToFetch.current > 0
  }, [fetchingTemplateActivityLogs, offsetToFetch.current])

  const isError = useCallback(() => {
    return !fetchingTemplateActivityLogs && templateActivityLogsFetchError
  }, [fetchingTemplateActivityLogs, templateActivityLogsFetchError])

  const isEmptyContent = useCallback(() => {
    return !fetchingTemplateActivityLogs && !templateActivityLogsFetchError && isEmpty(templateActivityLogs)
  }, [fetchingTemplateActivityLogs, templateActivityLogsFetchError, templateActivityLogs])

  return (
    <Container height={'100%'} className={css.container}>
      <Layout.Vertical flex={{ align: 'center-center' }} height={'100%'}>
        {isError() && <PageError message={templateActivityLogsFetchError} />}
        {isfetchingLogsFirstTime() && <PageSpinner message={getString('templatesLibrary.fetchingActivityLogs')} />}
        {isEmptyContent() ? (
          <Heading level={2} font={{ weight: 'bold' }} color={Color.GREY_300}>
            {getString('templatesLibrary.noActivityLogs')}
          </Heading>
        ) : (
          <Container
            style={{ overflow: 'auto' }}
            width={'100%'}
            height={'100%'}
            padding={{ top: 'medium', bottom: 'medium', left: 'xxlarge' }}
          >
            <>
              {(templateActivityLogs as AuditEventDTO[])?.map((auditEvent, index) => (
                <div ref={attachRefToLastElement(index) ? loadMoreRef : undefined} key={auditEvent.auditId}>
                  <TemplateActivity auditEvent={auditEvent} last={index === templateActivityLogs.length - 1} />
                </div>
              ))}
              {isfetchingLogsNextTime() && (
                <Container padding={{ left: 'xxlarge' }}>
                  <Text icon="loading" iconProps={{ size: 20 }} font={{ align: 'center' }}>
                    {getString('templatesLibrary.fetchingActivityLogs')}
                  </Text>
                </Container>
              )}
            </>
          </Container>
        )}
      </Layout.Vertical>
    </Container>
  )
}
