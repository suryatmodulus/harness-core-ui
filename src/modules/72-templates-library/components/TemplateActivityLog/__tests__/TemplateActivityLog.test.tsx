/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import { mockTemplates } from '@templates-library/TemplatesTestHelper'
import * as infiniteScrollHook from '../InfiniteScroll'
import { TemplateActivityLog } from '../TemplateActivityLog'
import { mockActivityLogs } from './TemplateActivityLogTestHelper'

describe('API ERROR', () => {
  test('if error message is displayed', () => {
    jest.spyOn(infiniteScrollHook, 'useInfiniteScroll').mockReturnValue({
      items: [],
      error: 'someerror',
      fetching: false,
      attachRefToLastElement: jest.fn(),
      hasMore: { current: false },
      loadItems: jest.fn(),
      offsetToFetch: { current: 0 }
    })
    const { queryByText } = render(<TemplateActivityLog template={mockTemplates?.data?.content?.[0] || {}} />)

    expect(queryByText('someerror')).toBeTruthy()
  })
})

describe('API FETCHING', () => {
  test('if loading message is displayed', () => {
    jest.spyOn(infiniteScrollHook, 'useInfiniteScroll').mockReturnValue({
      items: [],
      error: '',
      fetching: true,
      attachRefToLastElement: jest.fn(),
      hasMore: { current: false },
      loadItems: jest.fn(),
      offsetToFetch: { current: 0 }
    })
    const { queryByText } = render(
      <TestWrapper>
        <TemplateActivityLog template={mockTemplates?.data?.content?.[0] || {}} />
      </TestWrapper>
    )

    expect(queryByText('templatesLibrary.fetchingActivityLogs')).toBeTruthy()
  })
})

describe('API SUCCESS', () => {
  test('if content is displayed', () => {
    jest.spyOn(infiniteScrollHook, 'useInfiniteScroll').mockReturnValue({
      items: mockActivityLogs,
      error: '',
      fetching: false,
      attachRefToLastElement: jest.fn(),
      hasMore: { current: false },
      loadItems: jest.fn(),
      offsetToFetch: { current: 0 }
    })

    const { container } = render(
      <TestWrapper>
        <TemplateActivityLog template={mockTemplates?.data?.content?.[0] || {}} />
      </TestWrapper>
    )

    expect(container).toMatchSnapshot('response data')
  })

  test('if content is displayed and next batch is fetching', () => {
    jest.spyOn(infiniteScrollHook, 'useInfiniteScroll').mockReturnValue({
      items: mockActivityLogs,
      error: '',
      fetching: true,
      attachRefToLastElement: jest.fn(),
      hasMore: { current: false },
      loadItems: jest.fn(),
      offsetToFetch: { current: 1 }
    })

    const { queryByText } = render(
      <TestWrapper>
        <TemplateActivityLog template={mockTemplates?.data?.content?.[0] || {}} />
      </TestWrapper>
    )

    expect(queryByText('templatesLibrary.fetchingActivityLogs')).toBeTruthy()
  })

  test('if empty content is displayed', () => {
    jest.spyOn(infiniteScrollHook, 'useInfiniteScroll').mockReturnValue({
      items: [],
      error: '',
      fetching: false,
      attachRefToLastElement: jest.fn(),
      hasMore: { current: false },
      loadItems: jest.fn(),
      offsetToFetch: { current: 0 }
    })

    const { queryByText } = render(
      <TestWrapper>
        <TemplateActivityLog template={mockTemplates?.data?.content?.[0] || {}} />
      </TestWrapper>
    )

    expect(queryByText('templatesLibrary.noActivityLogs')).toBeTruthy()
  })
})
