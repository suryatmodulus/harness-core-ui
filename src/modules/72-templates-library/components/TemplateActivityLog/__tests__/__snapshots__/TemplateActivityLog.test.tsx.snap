// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API SUCCESS if content is displayed: response data 1`] = `
<div>
  <div
    class="StyledProps--font StyledProps--main container"
    style="height: 100%;"
  >
    <div
      class="Layout--vertical StyledProps--main StyledProps--flex StyledProps--flex-align-center-center"
      style="height: 100%;"
    >
      
      <div
        class="StyledProps--font StyledProps--main StyledProps--padding-top-medium StyledProps--padding-bottom-medium StyledProps--padding-left-xxlarge"
        style="overflow: auto; width: 100%; height: 100%;"
      >
        <div>
          <div
            class="activity"
          >
            <div>
              <button
                class="bp3-button Button--button StyledProps--font StyledProps--main timelineFirstIcon Button--with-current-color Button--round Button--iconOnly Button--without-shadow Button--withLeftIcon"
                type="button"
              >
                <span
                  class="bp3-icon bp3-icon-edit StyledProps--main"
                  icon="edit"
                >
                  <svg
                    data-icon="edit"
                    height="10"
                    viewBox="0 0 16 16"
                    width="10"
                  >
                    <desc>
                      edit
                    </desc>
                    <path
                      d="M3.25 10.26l2.47 2.47 6.69-6.69-2.46-2.48-6.7 6.7zM.99 14.99l3.86-1.39-2.46-2.44-1.4 3.83zm12.25-14c-.48 0-.92.2-1.24.51l-1.44 1.44 2.47 2.47 1.44-1.44c.32-.32.51-.75.51-1.24.01-.95-.77-1.74-1.74-1.74z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
              </button>
              <div
                class="timelineSeparator"
              />
              <div
                class="dot"
              />
              <div
                class="timelineSeparator"
              />
            </div>
            <div>
              <div
                class="Layout--horizontal Layout--layout-spacing-xsmall StyledProps--main StyledProps--margin-left-small StyledProps--margin-top-xsmall"
              >
                <p
                  class="StyledProps--font StyledProps--main StyledProps--intent-primary"
                >
                  Admin
                </p>
                <p
                  class="StyledProps--font StyledProps--main"
                >
                  templatesLibrary.default
                </p>
                <p
                  class="StyledProps--font StyledProps--main StyledProps--intent-primary"
                >
                  v2
                </p>
              </div>
              <p
                class="StyledProps--font StyledProps--main StyledProps--font-size-small StyledProps--margin-left-medium"
              >
                dummy date
              </p>
              <div
                class="StyledProps--font StyledProps--main StyledProps--padding-small StyledProps--margin-left-medium StyledProps--margin-top-medium StyledProps--border"
              >
                <p
                  class="StyledProps--font Text--lineclamp StyledProps--main StyledProps--font-size-small"
                  style="--text-line-clamp: 2; width: 420px;"
                >
                  someauditactivity
                </p>
              </div>
            </div>
          </div>
        </div>
        <div>
          <div
            class="activity"
          >
            <div>
              <button
                class="bp3-button Button--button StyledProps--font StyledProps--main timelineFirstIcon Button--with-current-color Button--round Button--iconOnly Button--without-shadow Button--withLeftIcon"
                type="button"
              >
                <span
                  class="bp3-icon bp3-icon-edit StyledProps--main"
                  icon="edit"
                >
                  <svg
                    data-icon="edit"
                    height="10"
                    viewBox="0 0 16 16"
                    width="10"
                  >
                    <desc>
                      edit
                    </desc>
                    <path
                      d="M3.25 10.26l2.47 2.47 6.69-6.69-2.46-2.48-6.7 6.7zM.99 14.99l3.86-1.39-2.46-2.44-1.4 3.83zm12.25-14c-.48 0-.92.2-1.24.51l-1.44 1.44 2.47 2.47 1.44-1.44c.32-.32.51-.75.51-1.24.01-.95-.77-1.74-1.74-1.74z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
              </button>
              <div
                class="timelineSeparator"
              />
              <div
                class="dot"
              />
              <div
                class="timelineSeparator"
              />
            </div>
            <div>
              <div
                class="Layout--horizontal Layout--layout-spacing-xsmall StyledProps--main StyledProps--margin-left-small StyledProps--margin-top-xsmall"
              >
                <p
                  class="StyledProps--font StyledProps--main StyledProps--intent-primary"
                >
                  common.userLabel
                </p>
                <p
                  class="StyledProps--font StyledProps--main"
                >
                  templatesLibrary.updated
                </p>
                <p
                  class="StyledProps--font StyledProps--main StyledProps--intent-primary"
                >
                  v2
                </p>
              </div>
              <p
                class="StyledProps--font StyledProps--main StyledProps--font-size-small StyledProps--margin-left-medium"
              >
                dummy date
              </p>
              <div
                class="StyledProps--font StyledProps--main StyledProps--padding-small StyledProps--margin-left-medium StyledProps--margin-top-medium StyledProps--border"
              >
                <p
                  class="StyledProps--font Text--lineclamp StyledProps--main StyledProps--font-size-small"
                  style="--text-line-clamp: 2; width: 420px;"
                >
                  templatesLibrary.sampleTemplateAuditComment
                </p>
              </div>
            </div>
          </div>
        </div>
        <div>
          <div
            class="activity"
          >
            <div>
              <button
                class="bp3-button Button--button StyledProps--font StyledProps--main timelineFirstIcon Button--with-current-color Button--round Button--iconOnly Button--without-shadow Button--withLeftIcon"
                type="button"
              >
                <span
                  class="bp3-icon bp3-icon-edit StyledProps--main"
                  icon="edit"
                >
                  <svg
                    data-icon="edit"
                    height="10"
                    viewBox="0 0 16 16"
                    width="10"
                  >
                    <desc>
                      edit
                    </desc>
                    <path
                      d="M3.25 10.26l2.47 2.47 6.69-6.69-2.46-2.48-6.7 6.7zM.99 14.99l3.86-1.39-2.46-2.44-1.4 3.83zm12.25-14c-.48 0-.92.2-1.24.51l-1.44 1.44 2.47 2.47 1.44-1.44c.32-.32.51-.75.51-1.24.01-.95-.77-1.74-1.74-1.74z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
              </button>
              <div
                class="timelineSeparator"
              />
              <div
                class="dot"
              />
              <div
                class="timelineSeparator timelineSeparatorLast"
              />
              <span
                class="bp3-icon bp3-icon-double-chevron-down StyledProps--main lastIcon StyledProps--color StyledProps--color-grey250"
                icon="double-chevron-down"
              >
                <svg
                  data-icon="double-chevron-down"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                >
                  <desc>
                    double-chevron-down
                  </desc>
                  <path
                    d="M7.29 8.71c.18.18.43.29.71.29s.53-.11.71-.29l4-4a1.003 1.003 0 00-1.42-1.42L8 6.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42l4 4zM12 8c-.28 0-.53.11-.71.29L8 11.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42l4 4c.18.18.43.29.71.29s.53-.11.71-.29l4-4A1.003 1.003 0 0012 8z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
            </div>
            <div>
              <div
                class="Layout--horizontal Layout--layout-spacing-xsmall StyledProps--main StyledProps--margin-left-small StyledProps--margin-top-xsmall"
              >
                <p
                  class="StyledProps--font StyledProps--main StyledProps--intent-primary"
                >
                  Admin
                </p>
                <p
                  class="StyledProps--font StyledProps--main"
                >
                  templatesLibrary.created
                </p>
                <p
                  class="StyledProps--font StyledProps--main StyledProps--intent-primary"
                >
                  v1
                </p>
              </div>
              <p
                class="StyledProps--font StyledProps--main StyledProps--font-size-small StyledProps--margin-left-medium"
              >
                dummy date
              </p>
              <div
                class="StyledProps--font StyledProps--main StyledProps--padding-small StyledProps--margin-left-medium StyledProps--margin-top-medium StyledProps--border"
              >
                <p
                  class="StyledProps--font Text--lineclamp StyledProps--main StyledProps--font-size-small"
                  style="--text-line-clamp: 2; width: 420px;"
                >
                  templatesLibrary.sampleTemplateAuditComment
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
